docs/hugo/content/design/ADR-2023-05-Installing-Only-Selected-CRDs.md

```diff
@@ -0,0 +1,174 @@
+
+* Addition of new resources: `kubectl patch configmap managed-resources -n azureserviceoperator-system --patch '{"data":{"microsoft.storage":"*"}}'`
+* Removal of old resources: `kubectl patch configmap managed-resources -n azureserviceoperator-system --type=json -p='[{"op": "remove", "path": "/data/microsoft.resources"}]'`
+
+If nothing is specified as part of the upgrade, the set of supported resources will not change.
+
+If the `ConfigMap` doesn't exist when we go to upgrade, we can assume we're coming from an older ASO version that didn't
+support the `ConfigMap` and default to creating one that signals to install all resources.
+
+Pros:
+* Meets all the requirements.
+* Works for both Helm and raw-YAML situations. Helm as discussed above and raw YAML by having users
+
+Cons:
+* Complicated to implement.
+* Not the behavior that Helm users would necessarily expect for Helm values.
+
+Open questions:
+* Do we even need to support a Helm value for configuring this? Technically speaking the only reason we would need to is
+  to enable ASO as an Extension or Addon, as there everything must be done through Helm. For users of Helm itself we likely
+  can just tell them to run a `kubectl` command afterwards to create the `ConfigMap`. That does make the Helm install 
+  experience worse though.
+* We may be able to create the `ConfigMap` as empty given Helms use of
+  [3 way strategic merge](https://helm.sh/docs/faq/changes_since_helm2/#improved-upgrade-strategy-3-way-strategic-merge-patches)
+  and then patch it afterwards.
+* Is it OK that _providers_ would be patchable, but you would need to include all the resources when patching something 
+  like `microsoft.storage: "StorageAccount,StorageAccountsBlobService"`? We could just disallow individual resource selection
+  for now, although that might eventually become a problem for certain large providers (`Microsoft.Network` comes to mind).
+
+See also Helm [post rendering](https://helm.sh/docs/topics/advanced/#post-rendering), which we could also consider.
+pre-install hook (or post-install) seems easier though.
+
+#### Answer 2c: Using a CRD
+
+We investigated doing this originally for [Helm chart size limitations](../ADR-2023-02-Helm-Chart-Size-Limitations)
+
+Pros:
+* Slightly more structured than a `ConfigMap`.
+
+Cons:
+* Chicken and egg problem of using the CRD that's part of the operator to configure the operator behavior. 
+  Helm has a very hard time creating a CRD _and_ an instance of the CRD in the same deployment.
+* May need to implement collection strategic merge behavior that already exists for `ConfigMap`.
+
+## Decision
+
+For question #1: [Answer 1c: Selection of CRDs to install is required and the operator pod will not launch without it](#answer-1c--selection-of-crds-to-install-is-required-and-the-operator-pod-will-not-launch-without-it)
+For question #2: [Answer 2b: Using a ConfigMap](#answer-2b--using-a-configmap)
```

Comment:
With the separation of question 2 into "how to configure" and "how to ensure consistency" (as mentioned elsewhere), I think we take a different approach:

* Mandatory configuration
* Retain Command line parameter
* Always support CRDs that have already been installed

Reply:
Not sure exactly what "Retain command line parameter" means - does it just mean "use the option we have now" or are you talking about some cmdline parameter that is retained between deployments somehow automatically?

I agree with your argument towards "Always support CRDs that have already been installed", but I think that actually forces us away from "Mandatory configuration", at least in the sense of Helm required parameters or anything like that.

I think what we want is:
- Operator exits with a clear reason (visible on Pod status) if no CRDs are configured.
- Cmdline parameter to configure which _new_ CRDs to install
- All existing CRDs are automatically upgraded to the latest version, even if not specified in cmdline parameter.

If we have some form of mandatory configuration _and_ the "existing CRDs" behavior it's not clear what the user should put into the mandatory configuration if they want to just use the resources they were already using.